### PR TITLE
[FLINK-20857][table-planner-blink] Separate the implementation of batch window aggregate nodes

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Separate the implementation of stream window aggregate nodes, including 
BatchExecHashWindowAggregate
BatchExecLocalHashWindowAggregate
BatchExecSortWindowAggregate
BatchExecLocalSortWindowAggregate
BatchExecPythonGroupWindowAggregate*

## Brief change log

  - *Rename BatchExecWindowAggregateBase to BatchPhysicalWindowAggregateBase and do some refactoring*
  - *Introduce BatchPhysicalHashWindowAggregate & BatchPhysicalLocalHashWindowAggregate, and make BatchExecHashWindowAggregate only extended from ExecNode*
  - *Introduce BatchPhysicalSortWindowAggregate & BatchPhysicalLocalSortWindowAggregate, and make BatchExecSortWindowAggregate only extended from ExecNode*
  - *Introduce BatchPhysicalPythonGroupWindowAggregate, and make BatchExecPythonGroupWindowAggregate only extended from ExecNode*

## Verifying this change

This change is a refactoring rework covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+/** Batch {@link ExecNode} for hash-based window aggregate operator. */
+public class BatchExecHashWindowAggregate extends ExecNodeBase<RowData>
+        implements BatchExecNode<RowData> {
+
+    private final int[] grouping;
+    private final int[] auxGrouping;
+    private final AggregateCall[] aggCalls;
+    private final LogicalWindow window;
+    private final int inputTimeFieldIndex;
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java:59]: the same as Stream, we can get from LogicalWindow?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    private final boolean inputTimeIsDate;
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java:60]: the same as inputTimeFieldIndex

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                hashWindowCodeGenerator.gen(
+                        inputRowType,
+                        (RowType) getOutputType(),
+                        groupBufferLimitSize,
+                        0,
+                        windowSizeAndSlideSize.f0,
+                        windowSizeAndSlideSize.f1);
+
+        final long managedMemory =
+                MemorySize.parse(
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java:142]: use the utility method

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortWindowAggregate.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+/** Batch {@link ExecNode} for sort-based window aggregate operator. */
+public class BatchExecSortWindowAggregate extends ExecNodeBase<RowData>
+        implements BatchExecNode<RowData> {
+
+    private final int[] grouping;
+    private final int[] auxGrouping;
+    private final AggregateCall[] aggCalls;
+    private final LogicalWindow window;
+    private final int inputTimeFieldIndex;
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortWindowAggregate.java:58]: ditto

**Review by wenlong88 — COMMENTED** [review]
the overall change LGTM, just left a minor comment.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java:58]: Different from `StreamExecWindowAggregate`, `BatchExecWindowAggregate` has local agg and global agg, and the values of  the `inputTimeFieldIndex ` are different for them. If we remove this parameter, we have to derive the value both in `BatchPhysicalWindowAggregateRule` and `BatchExecWindowAggregate`.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java (hunk 3 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecHashWindowAggregate.java]: done

**Review by wenlong88 — APPROVED** [review]
thanks for the explanation, LGTM for now.

**Comment by godfreyhe** [comment]
Thanks for the review, I will resolve the conflict in my local and merge the pr

**Pull request closed without merging**